File: python3/history.py
import abc


class History(abc.ABC):
    """Stores actions, states, rewards, and info about the env."""

    def __init__(self):
        self.a = []
        self.s = []
        self.r = []
        self.info = []

    def append(self, a: int, s: int, r: float, info: str):
        """Appends a, s, r and info."""
        self.a.append(a)
        self.s.append(s)
        self.r.append(r)
        self.info.append(info)

    def save(self, path: str):
        """Saves history into path."""
        with open(path, mode="w") as f:
            for a, s, r, info in zip(self.a, self.s, self.r, self.info):
                f.write("{}\t{}\t{:.15f}\t{}\n".format(a, s, r, info))

File: rust/src/history.rs
use crate::{Action, Reward, State};
use std::fs::File;
use std::io::BufWriter;
use std::io::Write;

pub struct History {
    pub a: Vec<Action>,
    pub s: Vec<State>,
    pub r: Vec<Reward>,
    pub info: Vec<String>,
}

impl History {
    pub fn new() -> Self {
        History {
            a: vec![],
            s: vec![],
            r: vec![],
            info: vec![],
        }
    }

    pub fn push(&mut self, a: Action, s: State, r: Reward, info: &str) {
        self.a.push(a);
        self.s.push(s);
        self.r.push(r);
        self.info.push(String::from(info));
    }

    pub fn save(&self, path: &str) -> std::io::Result<()> {
        let file = File::create(path)?;
        let mut writer = BufWriter::new(&file);

        for i in 0..self.a.len() {
            writer.write(
                format!(
                    "{}\t{}\t{:.15}\t{}\n",
                    self.a[i], self.s[i], self.r[i], &self.info[i]
                )
                .as_bytes(),
            )?;
        }

        writer.flush()?;

        Ok(())
    }
}
